tools/image_database_browser/image_database_browser.cc

```diff
@@ -26,35 +28,56 @@ int bobMain(int argc, char **argv)
 
     // Load images
     std::cout << "Loading images..." << std::endl;
-    const auto images = database.getImages();
+    std::vector<cv::Mat> images{ database.size() };
 
     // Iterate through images with arrow keys
     cv::namedWindow(argv[0]);
     for (size_t i = 0; i < database.size();) {
-        std::stringstream ss;
+        std::stringstream ssTitle, ssNumber;
         const auto pos = database[i].position;
-        ss << database.getName() << " [" << 1 + i << "/" << database.size() << "]"
-           << " (" << pos[0] << ", " << pos[1] << ", " << pos[2] << ")";
+        ssTitle << database.getName() << " [" << i << "/" << database.size() << "]"
+                << " (" << pos[0] << ", " << pos[1] << ", " << pos[2] << ")";
+        if (images[i].empty()) {
+            images[i] = database[i].load();
+        }
         cv::imshow(argv[0], images[i]);
-        cv::setWindowTitle(argv[0], ss.str());
+        cv::setWindowTitle(argv[0], ssTitle.str());
+
+        size_t newi = i;
         do {
-            switch (cv::waitKeyEx(50) & OS::KeyMask) {
+            const auto key = cv::waitKeyEx(50) & OS::KeyMask;
+            if (key >= '0' && key <= '9') {
+                ssNumber << static_cast<char>(key);
```

Comment:
What's the ``ssNumber`` logic doing? it's hard to tell

Reply:
Hmm I didn't really comment this terribly well because I expected that I would probably be the only person using it :wink:, but I agree that the logic's a bit opaque here. The idea is that if you want to jump to e.g. frame 100, you can do that by entering: 1, 0, 0, g. ``ssNumber`` just keeps a record of the last number keys that were pressed.

I've put in a comment to make this more obvious.

Reply:
That is definitely smarter than I assumed it was :smile: